Add new mo bills scraper

braykuka opened the pull request:
## Description
I've written senate and house scrapers in mo-bills.py. 

## Run
python -m scrapy crawl mo-bills -a session=2023 -O mo-bills.json -s "DOWNLOADER_MIDDLEWARES={}"
python -m scrapy crawl mo-bills -a session=2023 -a chamber=lower -O mo-bills.json -s "DOWNLOADER_MIDDLEWARES={}"
python -m scrapy crawl mo-bills -a session=2023 -a chamber=upper -O mo-bills.json -s "DOWNLOADER_MIDDLEWARES={}"
python -m scrapy crawl mo-bills -a session=2022S1 -O mo-bills.json -s "DOWNLOADER_MIDDLEWARES={}"
... ...

braykuka commented:
[mo-bills-result.zip](https://github.com/openstates/scrapy-test/files/13436078/mo-bills-result.zip)
Attached the full result.

jessemortenson commented on scrapers/scrapers/spiders/mo-bills.py:
```diff
@@ -0,0 +1,687 @@
+
+        if self.chamber in [Chamber.UPPER, None]:
+            # scrape senate bills
+            sb_url = 'https://www.senate.mo.gov/23info/BTS_Web/BillList.aspx?SessionType=R'
```
The session value in this URL is hard-coded here. Please add support for using the `session` parameter to construct `sb_url` like the original scraper: https://github.com/openstates/openstates-scrapers/blob/74648a9e6091c7e1f5e70c15572b4278bb2fdd71/scrapers/mo/bills.py#L655

braykuka replied:
@jessemortenson I've fixed this issue.

jessemortenson commented on scrapers/scrapers/spiders/mo-bills.py:
```diff
@@ -0,0 +1,687 @@
+            bl_response = Selector(text=bill_list_content, type='xml')
+
+            for bill in bl_response.xpath('//BillXML'):
+                bill_url = bill.xpath('./BillXMLLink/text()').get()
```
This section (173-178) looks like it is repeated below in `parse_house_bill_list()` (lines 193-198). Could `parse_house_bill_list()` be used fore both the "first" (unzipped) session and for an older (zipped) session? If so I think that would be preferable. It is a common mistake when maintaining scrapers to accidentally fix a selector/string in one section, and not realize it is repeated somewhere else.

braykuka replied:
@jessemortenson I've updated this issue.

jessemortenson commented on scrapers/scrapers/spiders/mo-bills.py:
```diff
@@ -0,0 +1,687 @@
+                rc_present = action.xpath(
+                    './RollCall/TotalPresent/text()').get()
+
+                #     vote = VoteEvent(
```
Noticing this is commented out. Did you run into problems with this code? This is primarily a Bills scraper, but there are cases where Vote data is mixed in with Bill data (on the source website), so it's good to yield the VoteEvent items as well.

braykuka replied:
Regarding this issue, I observed that the old scraper produces the VoteEvent items. However, if we choose to yield this vote in Scrapy, it will be in the same file as an item. If you prefer, we can uncomment this part.

jessemortenson commented on scrapers/scrapers/spiders/mo-bills.py:
```diff
@@ -0,0 +1,687 @@
+
+
+# remove whitespace, linebreaks, and end parentheses
+def clean_text(text):
```
This looks genuinely like a general utility, so would be a good candidate for a utils.py file

braykuka replied:
I've fixed it.

jessemortenson commented on scrapers/scrapers/spiders/mo-bills.py:
```diff
@@ -0,0 +1,687 @@
+                    on_duplicate="ignore",
+                )
+
+        # cosponsors show up on their own page, if they exist
```
Since we are chaining a series of requests (first to get co-sponsors, then to get actions, then to get versions, then to get fiscal notes, then finally to yield the bill), I think would be worth adding a comment here that describes this chain of requests that is going to happen, just to make it clearer to other devs

braykuka replied:
I've fixed it.

jessemortenson reviewed (changes requested):
Thanks, this is a really good start. I have not had a chance to actually run the scraper and validate data yet. But I'd like to request a few changes that seem apparent from a review of the code.

Running the scraper and validating the data will be one of my next tasks, early next week. If you prefer to wait to make further changes until I provide feedback from that step, that's OK. It's up to you.

jessemortenson commented:
Thanks for the additional code changes, I can review those next. I just completed a round of data validation. Overall things look good, but there are a few problems that are not so obvious but are important to fix. Please take a look at these:

## Code is stripping too much description text from the description of an action in some cases.

Losing time/location from description ("Monday, April 24, 2023, 1:30 p.m., Senate Committee Room 2Senate Committee Room 2")

```
https://house.mo.gov/Bill.aspx?bill=HB345&year=2023&code=R
ERROR:root:Differences found for HB 345 (also: HB 454)
ERROR:root:{
    "values_changed": {
        "root['actions'][20]['description']": {
            "new_value": "Public Hearing Scheduled (S)",
            "old_value": "Public Hearing Scheduled (S) - Monday, April 24, 2023, 1:30 p.m., Senate Committee Room 2"
        },
        "root['actions'][22]['description']": {
            "new_value": "Executive Session Scheduled",
            "old_value": "Executive Session Scheduled - Monday, May 1, 2023, 2:30 p.m., Senate Committee Room 2"
```

Losing additional details from description ("HA 1, adopted"):

```
https://house.mo.gov/BillContent.aspx?bill=HJR66&year=2023&code=R&style=new
ERROR:root:Differences found for HJR 66 (also: HB 12)
ERROR:root:{
    "values_changed": {
        "root['actions'][13]['description']": {
            "new_value": "Perfected with Amendments (H)",
            "old_value": "Perfected with Amendments (H) - HA 1, adopted"
            
```
            
##  Missing the "Witnesses" document from the .documents list

```
https://house.mo.gov/Bill.aspx?bill=HB345&year=2023&code=R
ERROR:root:Differences found for HB 345 (also: HB 734, HB 1022)

        "root['documents'][0]": {
            "new_value": {
                "note": "Bill Summary (Introduced)",
                "links": [
                    {
                        "url": "https://documents.house.mo.gov/billtracking/bills231/sumpdf/HB0345I.pdf",
                        "media_type": "application/pdf"
                    }
                ],
                "date": "",
                "classification": ""
            },
            "old_value": {
                "note": "Bill Summary (Witnesses)",
                "links": [
                    {
                        "url": "https://documents.house.mo.gov/billtracking/bills231/witnesses/HB345Testimony.pdf",
                        "media_type": "application/pdf"
                    }
                ],
                "date": "",
                "classification": ""
            }
        }
```
        
## Incorrect parsing of "distributed amendments":

https://house.mo.gov/BillContent.aspx?bill=HB454&year=2023&code=R&style=new

New/incorrect value is:

```
{
    "note": "Amendment None Distributed",
    "links": [
      {
        "url": "https://documents.house.mo.gov/billtracking/bills231/amendpdf/1194H02.02H.pdf",
        "media_type": "application/pdf"
      },
      {
        "url": "https://documents.house.mo.gov/billtracking/bills231/amendpdf/1194H02.03H.pdf",
        "media_type": "application/pdf"
      },
      {
        "url": "https://documents.house.mo.gov/billtracking/bills231/amendpdf/1194H02.04H.pdf",
        "media_type": "application/pdf"
      },
      {
        "url": "https://documents.house.mo.gov/billtracking/bills231/amendpdf/1194H02.05H.pdf",
        "media_type": "application/pdf"
      }
    ],
    "date": "",
    "classification": ""
  }
```


Old/correct value is:

```
{
    "note": "Amendment 1194H02.05H (Distributed)",
    "links": [
      {
        "url": "https://documents.house.mo.gov/billtracking/bills231/amendpdf/1194H02.05H.pdf",
        "media_type": "application/pdf"
      }
    ],
    "date": "",
    "classification": ""
  },
  {
    "note": "Amendment 1194H02.04H (Distributed)",
    "links": [
      {
        "url": "https://documents.house.mo.gov/billtracking/bills231/amendpdf/1194H02.04H.pdf",
        "media_type": "application/pdf"
      }
    ],
    "date": "",
    "classification": ""
  },
  {
    "note": "Amendment 1194H02.03H (Distributed)",
    "links": [
      {
        "url": "https://documents.house.mo.gov/billtracking/bills231/amendpdf/1194H02.03H.pdf",
        "media_type": "application/pdf"
      }
    ],
    "date": "",
    "classification": ""
  },
  {
    "note": "Amendment 1194H02.02H (Distributed)",
    "links": [
      {
        "url": "https://documents.house.mo.gov/billtracking/bills231/amendpdf/1194H02.02H.pdf",
        "media_type": "application/pdf"
      }
    ],
    "date": "",
    "classification": ""
  }
```  
  
## URLs should be fully qualified instead of just "ShowAmendment.aspx?Id=532" - using `response.urljoin()` might work

```
 cat mo-bills-2023-full-1.json | jq -c '.[] | select(.identifier | contains("SB 378"))' | jq .versions
[
  {
    "note": "0415S.03F - SS Adopted, as amended",
    "links": [
      {
        "url": "ShowAmendment.aspx?Id=532",
        "media_type": "application/pdf"
      }
    ],
    "date": "",
    "classification": ""
  },
  {
    "note": "0415S03.01S - SA 1 SS Adopted",
    "links": [
      {
        "url": "ShowAmendment.aspx?Id=543",
        "media_type": "application/pdf"
      }
    ],
    "date": "",
    "classification": ""
  },
  {
    "note": "0415S03.02S - SA 2 SS Adopted",
    "links": [
      {
        "url": "ShowAmendment.aspx?Id=544",
        "media_type": "application/pdf"
      }
    ],
    "date": "",
    "classification": ""
  },
```

braykuka commented:
### Code is stripping too much description text from the description of an action in some cases.
I've fixed this issue by merging the separated elements within the XML, such as Description, Comments, and RollCall.

### Others
Additionally, I've addressed other issues. Changes on the document.house.mo.gov website required adjustments to specific code parts related to the House Bill, and these have been fixed.

jessemortenson reviewed (changes requested):
Thank you for the revision, that resolved the issues I saw. I would love one more round of changes, summary of comments:

- a couple very small function rename requests
- a couple sections of commented code that maybe was left in and unnecessary
- please uncomment the Vote Event logic, so we yield the Vote Events as well

braykuka commented:
@jessemortenson Thanks for your review. I've addressed the issues you mentioned. Please review them and let me know your opinion.
